File: AUG 2024/Rotate a Linked List.java
class Solution {
    // Function to rotate a linked list.
    public Node rotate(Node head, int k) {
        // add code here
        if(head == null && head.next == null){
            return head;
        }
        int size =0;
        Node temp = head;
        Node last = null;
        while(temp != null){
            size++;
            last = temp;
            temp = temp.next;
        }
        
        Node curr = head;
        Node next = head.next;
        while(k > 0){
            curr.next = null;
            last.next = curr;
            last = last.next;
            curr = next;
            next = curr.next;
            k--;
        }
        return curr;
    }
}

File: JAN 2025/Rotate a Linked List.py
class Solution:
    
    #Function to rotate a linked list.
    def rotate(self, head, k):
        # code here
        # contrain is that there is atleast 1 node, but k can be 0 
        if k == 0:
            return head
        # get the length of the linked list to handle large k using k modulo length
        curr = head
        l = 1
        while curr.next is not None:
            curr = curr.next
            l += 1
        k %= l
        # if k is multiple of length, then no change in the linked list
        if k == 0:
            return head
        # make the current linkedlist circular
        curr.next = head
        # Traverse the linked list to find the kth node
        curr = head
        for _ in range(1, k):
            curr = curr.next
        # Update the (k + 1)th node as the new head
        head = curr.next
        # Break the loop by updating the next pointer of kth node
        curr.next = None
        return head
